pkg/vm/engine/tae/logtail/handle.go

```diff
@@ -342,29 +347,71 @@ func (b *TableLogtailRespBuilder) Close() {
 	}
 }
 
-func (b *TableLogtailRespBuilder) visitBlkMeta(e *catalog.BlockEntry) {
+func (b *TableLogtailRespBuilder) visitBlkMeta(e *catalog.BlockEntry) (skipData bool) {
+	var latestCommittedNode *catalog.MetadataMVCCNode
+	newEnd := b.end
 	e.RLock()
-	mvccNodes := e.ClonePreparedInRange(b.start, b.end)
+	// try to find new end
+	if newest := e.GetLatestCommittedNode(); newest != nil {
+		latestCommittedNode = newest.CloneAll().(*catalog.MetadataMVCCNode)
+		latestPrepareTs := latestCommittedNode.GetPrepare()
+		if latestPrepareTs.Greater(b.end) {
+			newEnd = latestPrepareTs
+		}
+	}
+	mvccNodes := e.ClonePreparedInRange(b.start, newEnd)
 	e.RUnlock()
+
 	for _, node := range mvccNodes {
 		metaNode := node.(*catalog.MetadataMVCCNode)
-		var dstBatch *containers.Batch
-		if !metaNode.HasDropCommitted() {
-			dstBatch = b.blkMetaInsBatch
-			dstBatch.GetVectorByName(pkgcatalog.BlockMeta_ID).Append(e.ID)
-			dstBatch.GetVectorByName(pkgcatalog.BlockMeta_EntryState).Append(e.IsAppendable())
-			dstBatch.GetVectorByName(pkgcatalog.BlockMeta_CreateAt).Append(metaNode.CreatedAt)
-			dstBatch.GetVectorByName(pkgcatalog.BlockMeta_DeleteAt).Append(metaNode.DeletedAt)
-			dstBatch.GetVectorByName(pkgcatalog.BlockMeta_MetaLoc).Append([]byte(metaNode.MetaLoc))
-			dstBatch.GetVectorByName(pkgcatalog.BlockMeta_DeltaLoc).Append([]byte(metaNode.DeltaLoc))
-		} else {
-			dstBatch = b.blkMetaDelBatch
-		}
+		b.appendBlkMeta(e, metaNode)
+	}
 
-		dstBatch.GetVectorByName(catalog.AttrRowID).Append(u64ToRowID(e.ID))
-		dstBatch.GetVectorByName(catalog.AttrCommitTs).Append(metaNode.GetEnd())
-		dstBatch.GetVectorByName(catalog.AttrAborted).Append(metaNode.IsAborted())
+	if e.IsAppendable() {
+		if latestCommittedNode.MetaLoc != "" {
+			// appendable block has been flushed, no need to collect data
+			return true
+		}
+	} else {
+		if latestCommittedNode.DeltaLoc != "" && latestCommittedNode.GetEnd().GreaterEq(b.end) {
+			// non-appendable block has newer delta data on s3, no need to collect data
+			return true
+		}
 	}
+	return false
+}
+
+func (b *TableLogtailRespBuilder) appendBlkMeta(e *catalog.BlockEntry, metaNode *catalog.MetadataMVCCNode) {
+	// var dstBatch *containers.Batch
```

Comment:
What is the purpose of this comment?   Please either add a real, clear message on what other people need to know, or remove dead code.